### ISSUE TITLE
Why do results differ for HNSW from index factory vs. direct init?

### ISSUE THREAD

**Issue by tholor** [facebookresearch/faiss#1434]
# Summary
Probably I am missing something obvious here, but how can I initialize HNSW32Flat from the index factory? 

Using 
`faiss.index_factory(d, "HNSW32", faiss.METRIC_INNER_PRODUCT)` 
gives different results than 
`index = faiss.IndexHNSWFlat(d, 32, faiss.METRIC_INNER_PRODUCT)`

# Platform

OS: Ubuntu 18.04
Faiss version: faiss-cpu==1.6.3

Running on:
- [x] CPU
- [ ] GPU

Interface: 
- [ ] C++
- [x] Python

# Reproduction instructions

```
k = 5
nq = 100
nb = 10000
d = 32

# Query vectors
xq = faiss.randn((nq, d), seed=11)

# Doc Vectors
xb1 = faiss.randn((nb, d), seed=12)

# Reference
index = faiss.IndexFlatIP(d)
index.add(xb1)
Dref, Iref = index.search(xq, k)

# Init HNSW directly
index = faiss.IndexHNSWFlat(d, 32, faiss.METRIC_INNER_PRODUCT)
index.add(xb1)
D, I = index.search(xq, k)
print("Similarity (direct): ", np.sum(I == Iref)/I.size)

# Init HNSW via Index Factory
index = faiss.index_factory(d, "HNSW32", faiss.METRIC_INNER_PRODUCT)
index.add(xb1)
D, I = index.search(xq, k)
print("Similarity (Factory): ", np.sum(I == Iref)/I.size)
# Results in Outputs: 
# Similarity (direct):  0.688
# Similarity (Factory):  0.012
``` 

BTW: Thanks for your awesome work :)

**Comment by mdouze** [comment]
I tried this on the current Faiss, and I can't repro. 
What version of Faiss are you using?

**Comment by tholor** [comment]
faiss-cpu 1.6.3

**Comment by mdouze** [comment]
Please compile from source. 
There is currently a problem that anything after the release of the 1.6.3 packages is reported as 1.6.3, so the __version__ is not accurate.

**Comment by tholor** [comment]
I now compiled from source (for cpu, Ubuntu 18.04) using latest master, but I still get the same results.
Also tried different conda and pip versions without success. 

I noticed that there's a difference in the attribute `metric_type` of the index after the two different inits: 
- Factory: metric_type = 1  
- direct: metric_type = 0

So to me, it seems that with the factory the inner product metric is not correctly set. Not sure why. 

I guess I will just continue with the direct init of the class for now.
